### ISSUE TITLE
twisted/mail/smtp.py SenderMixin error logging bug

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2023]
|<img alt="shieldszero's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| shieldszero reported|
|-|-|
|Trac ID|trac#2023|
|Type|defect|
|Created|2006-08-28 22:18:20Z|

Because of a typo, SenderMixin's sentMail() method reproduces the outer loop test in its inner loop.

Attachments:

* [smtp.py.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/445/445cd2fd3273962bdf09425109a2d09f7170e837/ffbf5f123b9f54a36e561cc6cab2b6d6a666850a.diff) (427 bytes) - added by shieldszero on 2006-08-28 22:18:30Z - 
* [mail-smtp-senderMixin-sentMail.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/445/445cd2fd3273962bdf09425109a2d09f7170e837/55752f0b629c5201cdd5aa0673d8b31e3480b079.patch) (1959 bytes) - added by jesstess on 2009-09-01 17:56:47Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2023 2023
type__defect defect
reporter__shieldszero shieldszero
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1156803500000000 1156803500000000
changetime__1251835519000000 1251835519000000
version__None None
owner__ 
cc__shieldszero cc__jesstess cc__thijs cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_2"></a>|@jesstess removed owner|
|-|-|

I agree that it looks like the intention was to only log lines for addresses with unsuccessful SMTP response codes and that this is a typo.

The updated patch makes the one line fix in smtp.py and adds a sentMail unit test to test_smtp.py that checks that the SMTPDeliveryError passed to the factory errback doesn't contain addresses with successful SMTP response codes.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

Looks good to me doc-wise, ill leave the code-review up to someone else.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @jesstess|
|-|-|

There's a simpler way to do what `checkErr` is doing in this patch.  `Failure` (which is what the `deferred` argument will be an instance of, so that name is a bit misleading ;) has a `trap` method.  So you want to do something like this:

```python
def checkErr(self, failure):
    failure.trap(smtp.SMTPDeliveryError)
    self.assertEquals(failure.value.log, ...)
```

But there's an even easier way, too.  In the test method, you can use `TestCase.assertFailure`:

```python
onDone = self.assertFailure(defer.Deferred(), smtp.SMTPDeliveryError)
onDone.addCallback(lambda e: self.assertEquals(e.log, ...))

...
return onDone
```

Also, definitely make sure to add that `return onDone`.  Whenever you've got asynchronous stuff going on in a test, particularly when there are Deferreds involved, you are probably going to want to return a Deferred from the test method.  Otherwise parts of the test may not actually run, or a failure might get obscured or reported as part of the wrong test.

Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_5"></a>|@jesstess removed owner|
|-|-|

Thank you exarkun for the review; that is definitely a testing idiom for my proverbial back pocket. The new version of the patch has the better assertFailure way of testing the Exception.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Great.  I make a couple more minor tweaks - renamed a couple things and adjusted the style of the docstring.  I'll apply it now.  Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [27282]) Apply slightly modified mail-smtp-senderMixin-sentMail.patch

Author: jesstess, exarkun
Reviewer: thijs, exarkun
Fixes: [#2023](https://github.com/twisted/twisted/issues/2023)

Fix a logging bug in `twisted.mail.smtp.SenderMixin` which caused additional
information about non-failed recipients to be included in the error log about
failed recipients.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
